azure/datalake/store/lib.py

```diff
@@ -137,59 +133,60 @@ def auth(tenant_id=None, username=None,
     if not client_secret:
         client_secret = os.environ.get('azure_client_secret', None)
 
-    # You can explicitly authenticate with 2fa, or pass in nothing to the auth call
-    # and the user will be prompted to login interactively through a browser.
-
-    @retry_decorator_for_auth(retry_policy=retry_policy)
+    if client_secret:
+        contextClient = msal.ConfidentialClientApplication(client_id=client_id, authority=authority+tenant_id, client_credential=client_secret, http_cache=_http_cache)
+    else:
+        contextPub = msal.PublicClientApplication(client_id=client_id, authority=authority+tenant_id, http_cache=_http_cache)
+    
+    scopes = kwargs.get('scopes', ["https://datalake.azure.net/.default"])
     def get_token_internal():
-        # Internal function used so as to use retry decorator
         if require_2fa or (username is None and password is None and client_secret is None):
-            code = context.acquire_user_code(resource, client_id)
-            print(code['message'])
-            out = context.acquire_token_with_device_code(resource, code, client_id)
+            flow = contextPub.initiate_device_flow(scopes=scopes)
+            print(flow['message'])
+            out = contextPub.acquire_token_by_device_flow(flow)
         elif username and password:
-            out = context.acquire_token_with_username_password(resource, username,
-                                                               password, client_id)
+            out = contextPub.acquire_token_by_username_password(username=username, password=password, scopes=scopes)
         elif client_id and client_secret:
-            out = context.acquire_token_with_client_credentials(resource, client_id,
-                                                                client_secret)
+            out = contextClient.acquire_token_for_client(scopes=scopes)
             # for service principal, we store the secret in the credential object for use when refreshing.
             out.update({'secret': client_secret})
         else:
             raise ValueError("No authentication method found for credentials")
         return out
```

Comment:
While this new implementation will work in happy path, the separation of `PublicClientApplication` and `ConfidentialClientApplication` in MSAL happens to make the current implementation fragile. For example, if a user (accidentally?) configure `username` and `password` AND `client_secret`, the current implementation would run into a `contextPub` undefined exception.

Let's do some refactoring.

```python
scopes = kwargs.get('scopes', ["https://datalake.azure.net/.default"])
if client_secret:
    app = msal.ConfidentialClientApplication(
        client_id=client_id, authority=authority+tenant_id, client_credential=client_secret, http_cache=_http_cache)
    out = contextClient.acquire_token_for_client(scopes=scopes)
    # for service principal, we store the secret in the credential object for use when refreshing.
    out.update({'secret': client_secret})
else:
    app = msal.PublicClientApplication(client_id=client_id, authority=authority+tenant_id, http_cache=_http_cache)
    if require_2fa or (username is None and password is None):
        flow = app.initiate_device_flow(scopes=scopes)
        print(flow['message'])
        out = app.acquire_token_by_device_flow(flow)
    elif username and password:
        out = app.acquire_token_by_username_password(username=username, password=password, scopes=scopes)
    else:
        raise ValueError("No authentication method found for credentials")
```

Reply:
This does change the order though as previously if both client secret and require_2fa were specified, we would go the 2fa route. 

I am pushing a diff with some code duplication but it preserves the previous conditional behavior.

Reply:
Setting both client_secret and require_2fa was not a valid setup in the first place. And now we have to become backward compatible with the old implementation detail which happened to run the 2fa code path?

That being said, you raised a good point, given that this is an older library that will retire within a year. Lesson learned. We can go with your new commit.